livekit-plugins/livekit-plugins-assemblyai/livekit/plugins/assemblyai/stt.py

```diff
@@ -0,0 +1,395 @@
+
+        if not wait:
+            self._main_task.cancel()
+
+        with suppress(asyncio.CancelledError):
+            await self._main_task
+
+        await self._session.close()
+
+    @utils.log_exceptions(logger=logger)
+    async def _main_task(self) -> None:
+        await self._run(self._max_retry)
+
+    @utils.log_exceptions(logger=logger)
+    async def _run(self, max_retry: int) -> None:
+        """
+        Run a single websocket connection to AssemblyAI and make sure to reconnect
+        when something went wrong.
+        """
+        try:
+            retry_count = 0
+            while not self._closed:
+                try:
+                    live_config = {
+                        "sample_rate": self._opts.sample_rate,
+                        "word_boost": self._opts.word_boost,
+                        "encoding": self._opts.encoding,
+                        "disable_partial_transcripts": self._opts.disable_partial_transcripts,
+                        "enable_extra_session_information": self._opts.enable_extra_session_information,
+                    }
+
+                    headers = {
+                        "Authorization": self._api_key,
+                        "Content-Type": "application/json",
+                    }
+
+                    ws_url = "wss://api.assemblyai.com/v2/realtime/ws"
+                    filtered_config = {k: v for k, v in live_config.items() if v is not None}
+                    url = f"{ws_url}?{urlencode(filtered_config).lower()}"
+                    ws = await self._session.ws_connect(url, headers=headers)
+                    retry_count = 0  # connected successfully, reset the retry_count
+
+                    await self._run_ws(ws)
+                except Exception:
+                    # Something went wrong, retry the connection
+                    if retry_count >= max_retry:
+                        logger.error(
+                            f"failed to connect to AssemblyAI after {max_retry} tries"
+                        )
+                        break
+
+                    retry_delay = min(retry_count * 2, 10)  # max 10s
+                    retry_count += 1  # increment after calculating the delay, the first retry should happen directly
+
+                    logger.info(
+                        f"AssemblyAI connection failed, retrying in {retry_delay}s",
+                    )
+                    await asyncio.sleep(retry_delay)
+        finally:
+            self._event_queue.put_nowait(None)
+
+    async def _run_ws(self, ws: aiohttp.ClientWebSocketResponse) -> None:
+        """
+        This method can throw ws errors, these are handled inside the _run method
+        """
+
+        closing_ws = False
+        END_UTTERANCE_SILENCE_THRESHOLD_MSG = json.dumps(
+            {"end_utterance_silence_threshold": self._opts.end_utterance_silence_threshold }
+        )
+        self._queue.put_nowait(END_UTTERANCE_SILENCE_THRESHOLD_MSG)
```

Comment:
Only do this if the threshold is not None. Otherwise we can use the default settings from our API